### ISSUE TITLE
Overlapping serializer sources should be merged on writing

### ISSUE THREAD

**Issue by tbitai** [encode/django-rest-framework#6345]
## Checklist

- [x] I have verified that that issue exists against the `master` branch of Django REST framework.
- [x] I have searched for similar issues in both open and closed tickets and cannot find a duplicate.
- [x] This is not a usage question. (Those should be directed to the [discussion group](https://groups.google.com/forum/#!forum/django-rest-framework) instead.)
- [x] This cannot be dealt with as a third party library. (We prefer new functionality to be [in the form of third party libraries](https://www.django-rest-framework.org/topics/third-party-resources/#about-third-party-packages) where possible.)
- [x] I have reduced the issue to the simplest possible case.
- [x] I have included a failing test as a pull request. (If you are unable to do so we can still accept the issue.)

## Steps to reproduce

Create a serializer with two fields:

1. field with `source='parent'` and `to_internal_value()` return value `{'child1': data}` ("deep field") or `{'child1': data[0], 'child2': data[1]}` ("conflicting deep field"),
2. field with `source='parent.child2'` ("shallow field").

Call `is_valid()` and then check `validated_data`.

## Expected behavior

Data should be merged; in case of conflict, validation error should be raised.

## Actual behavior

The field defined later overrides the earlier ones.

**Comment by carltongibson** [comment]
This is not a use-case we want to address. The underlying `to_internal_value()`mechanism leverages a `set_value()` helper which deliberately mirrors Python's built-in `dict`'s behaviour: 

https://github.com/encode/django-rest-framework/blob/a52087b1b81f300d2a27dd262b452689fe195229/rest_framework/fields.py#L117-L135

Yes, later assignments clobber earlier ones. That's part of the design. 

For the use-case here, following @rpkilby's suggestions on the PR to restructure the serialisers is the way to go. Failing that a custom base serializer re-implementing `to_internal_value()` would be required.

**Comment by JPDSousa** [comment]
Joining late in the discussion 👋 

**tl;dr -** let me know if I should create a new issue if my use case doesn't properly intercept with this one.

---

# Problem statement

I've bumped into this limitation while trying to implement a serializer which represents a view from multiple models:

```python
class Person:
    name: str
    cars: CarStats
    stocks: StockStats

class WealthSerializer(serializers.Serializer):
    cars = serializers.FloatField(source='cars.total_price')
    stocks = serializers.FloatField(source='stocks.total_value')

class CarsSerializer(serializers.Serializer):
    total_cars = serializers.IntegerField(source='cars.count')
    oldest_car_date = serializers.DateField(source='cars.oldest_date')

class PersonSerializer(serializers.Serializer):
    name = serializers.CharField()
    wealth = WealthSerializer(source='*')
    cars = CarsSerializer(source='*')
```

Just like in the issue description, `cars` and `wealth` clash upon deserialization, as both update `Person.cars`.

# Proposed solution

> This is not a use-case we want to address. The underlying `to_internal_value()`mechanism leverages a `set_value()` helper which deliberately mirrors Python's built-in `dict`'s behaviour:
> 
> https://github.com/encode/django-rest-framework/blob/a52087b1b81f300d2a27dd262b452689fe195229/rest_framework/fields.py#L117-L135

I agree that supporting yet another corner case (i.e., first time bumping into this limitation in years) makes the code hard to read, hence my solution was to:
1. Replace `set_value` by an overridable instance method (also called `set_value`) in `Serializer.to_internal_value`
2. Subclass the (now patched) `Serializer` and override the new `set_value` instance method to perform a "deep update" when updating `ret`.

**Note:** While point 1. is mandatory in my proposal, point 2. can be implemented by the actual framework users ad-hoc.

# Discussion

Here's some pros/cons of the approach to jumpstart the discussion:

**Pros:**
* Making `set_value` overridable allows for custom conflict resolution policies other than the currently hardcoded [LWW](https://en.wikipedia.org/wiki/Eventual_consistency);
* The proposal (point 1. only) is clean and has no behavioural change.
* No breaking changes introduced

**Cons:**
* (inherited by any other point of extension) We're introducing one more way for users to mess up integrations.
* Note on breaking changes: introducing a new method might clash with user defined methods

WDYT?

**Comment by carltongibson** [comment]
> WDYT?

I think likely this (even just making the `set_value` hook overridable) is adding an extra complexity that's not worth it. 
I think asking you to implement `to_internal_value` is not too much to save that for this case. 

Likely you should raise this as a new discussion if you want to pursue it. 

Thanks 🙂

**Comment by auvipy** [comment]
> WDYT?

please open a discussion referencing this and what have you described.

**Comment by JPDSousa** [comment]
Done: https://github.com/encode/django-rest-framework/discussions/8782
